### PR TITLE
new(modern_probe): Add getuid, getgid, geteuid, getegid

### PR DESCRIPTION
**Pull request by dwindsor**
Signed-off-by: David Windsor <dwindsor@secureworks.com>

Add modern probes for:
- getuid
- getgid
- geteuid
- getegid

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test
 /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

/area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

> /area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

#723 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by Andreagit97** [comment]
wow man, you are on :fire: I will take a look ASAP!

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getuid.bpf.c [hunk 1]
```diff
@@ -14,19 +15,19 @@ int BPF_PROG(getuid_e,
 	     struct pt_regs *regs,
 	     long id)
 {
-	struct ringbuf_struct ringbuf;
```
Comment (hbrueckner) [inline driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getuid.bpf.c]: Based on discussion in [comment](https://github.com/falcosecurity/libs/pull/752#discussion_r1042446028) in #752, let's keep the ringbuf here. So you can skip this commit.

File: test/modern_bpf/test_suites/syscall_exit_suite/getuid_x.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+TEST(SyscallExit, getuidX)
+{
+	auto evt_test = get_syscall_event_test(__NR_getuid, EXIT_EVENT);
+
+	evt_test->enable_capture();
+
+	/*=============================== TRIGGER SYSCALL ===========================*/
+
+	assert_syscall_state(SYSCALL_SUCCESS, "getuid", syscall(__NR_getuid), NOT_EQUAL, -1);
+	uid_t uid = syscall(__NR_getuid);
```
Comment (hbrueckner) [inline test/modern_bpf/test_suites/syscall_exit_suite/getuid_x.cpp:13]: With this change:
````
uid_t uid = syscall(__NR_getuid);
assert_syscall_state(SYSCALL_SUCCESS, "getuid", uid, NOT_EQUAL, -1);
````
you could save one system call (similar approach you can find, for example, in `socket` test cases).

**Review by hbrueckner — CHANGES_REQUESTED** [review]
Thanks a lot @dwindsor!
And you are really on :fire: 

There are some proposed changes from #752 that can applied here as well. Also please add those events to the `event_prog_names` as outlined in [here](https://github.com/falcosecurity/libs/pull/752#discussion_r1042457192).

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getegid.bpf.c [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+#include <helpers/interfaces/variable_size_event.h>
+
+/*=============================== ENTER EVENT ===========================*/
+
+SEC("tp_btf/sys_enter")
+int BPF_PROG(getegid_e,
+	     struct pt_regs *regs,
+	     long id)
+{
+	struct auxiliary_map *auxmap = auxmap__get();
```
[the remark this answers is not in the document]
Comment (hbrueckner) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getegid.bpf.c:18]: Let me cite @Andreagit97 here:

> The rationale is that when the event has a known size at compile time we can use the ringbuf_ approach instead the auxmap_ one. The reason is that with the ringbuf we avoid a copy in memory so the approach is more efficient!

The system call path is always "critical" in sense of performance.  So when hooking into syscalls, let's try to be efficient as possible as calling BPF/kprobes/... programs slow down the system call path.

File: test/modern_bpf/test_suites/syscall_exit_suite/getuid_x.cpp (hunk 2 above)
Comment (dwindsor) [reply test/modern_bpf/test_suites/syscall_exit_suite/getuid_x.cpp:12]: Thanks! Fixed here as well as in getgid_x.cpp, geteuid_x.cpp and getegid_x.cpp.

**Review by Andreagit97 — COMMENTED** [review]
Only some minor details, after that we should be ready on my side. If you want to run `clang-format` as in the [old PR](https://github.com/falcosecurity/libs/pull/752#discussion_r1052416089) feel free to do that :)

**Comment by Andreagit97** [comment]
I know conflicts are quite a pain, unfortunately, we will have a conflict in the `events_prog_names.h` file every time we merge a set of syscalls, so be ready to do that again and again for all your PRs :/

**Comment by hbrueckner** [comment]
> Only some minor details, after that we should be ready on my side.

Also looked through, with the suggested changes from @Andreagit97 , this also looks fine to me.

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getuid.bpf.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2022 The Falco Authors.
+ *
+ * This file is dual licensed under either the MIT or GPL 2. See MIT.txt
+ * or GPL2.txt for full copies of the license.
+ */
+
+#include <helpers/interfaces/fixed_size_event.h>
+#include <helpers/interfaces/variable_size_event.h>
```
[the remark this answers is not in the document]
Comment (dwindsor) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getuid.bpf.c:9]: Fixed

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getgid.bpf.c [hunk 5]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2022 The Falco Authors.
+ *
+ * This file is dual licensed under either the MIT or GPL 2. See MIT.txt
+ * or GPL2.txt for full copies of the license.
+ */
+
+#include <helpers/interfaces/fixed_size_event.h>
+#include <helpers/interfaces/variable_size_event.h>
```
[the remark this answers is not in the document]
Comment (dwindsor) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getgid.bpf.c:9]: Fixed

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/geteuid.bpf.c [hunk 6]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2022 The Falco Authors.
+ *
+ * This file is dual licensed under either the MIT or GPL 2. See MIT.txt
+ * or GPL2.txt for full copies of the license.
+ */
+
+#include <helpers/interfaces/fixed_size_event.h>
+#include <helpers/interfaces/variable_size_event.h>
```
[the remark this answers is not in the document]
Comment (dwindsor) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/geteuid.bpf.c:9]: Fixed

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getegid.bpf.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2022 The Falco Authors.
+ *
+ * This file is dual licensed under either the MIT or GPL 2. See MIT.txt
+ * or GPL2.txt for full copies of the license.
+ */
+
+#include <helpers/interfaces/fixed_size_event.h>
+#include <helpers/interfaces/variable_size_event.h>
```
[the remark this answers is not in the document]
Comment (dwindsor) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getegid.bpf.c:9]: Fixed

File: test/modern_bpf/test_suites/syscall_exit_suite/getegid_x.cpp [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+		return;
+	}
+
+	evt_test->parse_event();
+
+	evt_test->assert_header();
+
+	/*=============================== ASSERT PARAMETERS  ===========================*/
+
+	/* Parameter 1: res (type: PT_UID) */
```
[the remark this answers is not in the document]
Comment (dwindsor) [reply test/modern_bpf/test_suites/syscall_exit_suite/getegid_x.cpp]: Fixed here as well as in the others.

File: test/drivers/test_suites/syscall_exit_suite/getegid_x.cpp [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+		return;
+	}
+
+	evt_test->parse_event();
+
+	evt_test->assert_header();
+
+	/*=============================== ASSERT PARAMETERS  ===========================*/
+
+	/* Parameter 1: egid (type: PT_UID) */
```
Comment (hbrueckner) [inline test/drivers/test_suites/syscall_exit_suite/getegid_x.cpp:32]: typo: should be of type `PT_GID`

File: test/drivers/test_suites/syscall_exit_suite/getgid_x.cpp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+		return;
+	}
+
+	evt_test->parse_event();
+
+	evt_test->assert_header();
+
+	/*=============================== ASSERT PARAMETERS  ===========================*/
+
+	/* Parameter 1: gid (type: PT_UID) */
```
Comment (hbrueckner) [inline test/drivers/test_suites/syscall_exit_suite/getgid_x.cpp:32]: typo: should be of type `PT_GID`

**Review by hbrueckner — APPROVED** [review]
Hi @dwindsor ,
I just have two more or less typos to be very correct in the comments which should not really block this PR moving forward.

Thank you very much for those 4 additional syscalls towards a complete modern BPF!
/approve

**Comment by Andreagit97** [comment]
yeah this should be ready! but we need to resolve conflicts :(

**Comment by dwindsor** [comment]
> yeah this should be ready! but we need to resolve conflicts :(

Rebased =)

File: test/drivers/test_suites/syscall_exit_suite/getegid_x.cpp (hunk 9 above)
Comment (dwindsor) [reply test/drivers/test_suites/syscall_exit_suite/getegid_x.cpp]: Fixed

File: test/drivers/test_suites/syscall_exit_suite/getgid_x.cpp (hunk 10 above)
Comment (dwindsor) [reply test/drivers/test_suites/syscall_exit_suite/getgid_x.cpp]: Fixed

**Review by Andreagit97 — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**